File: app/src/main/java/blue/koenig/kingsfinances/model/FinanceModel.java
package blue.koenig.kingsfinances.model;

import android.content.Context;
import android.graphics.Color;

import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.Category;
import com.koenig.commonModel.Component;
import com.koenig.commonModel.Item;
import com.koenig.commonModel.ItemType;
import com.koenig.commonModel.Operation;
import com.koenig.commonModel.Operator;
import com.koenig.commonModel.User;
import com.koenig.commonModel.database.DatabaseItem;
import com.koenig.commonModel.finance.Balance;
import com.koenig.commonModel.finance.BankAccount;
import com.koenig.commonModel.finance.Expenses;
import com.koenig.commonModel.finance.StandingOrder;
import com.koenig.communication.messages.AUDMessage;
import com.koenig.communication.messages.AskForUpdatesMessage;
import com.koenig.communication.messages.FamilyMessage;
import com.koenig.communication.messages.UpdatesMessage;
import com.koenig.communication.messages.finance.FinanceTextMessages;

import org.joda.time.DateTime;
import org.joda.time.Period;

import java.nio.ByteBuffer;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import blue.koenig.kingsfamilylibrary.model.FamilyConfig;
import blue.koenig.kingsfamilylibrary.model.communication.ServerConnection;
import blue.koenig.kingsfamilylibrary.model.family.FamilyModel;
import blue.koenig.kingsfamilylibrary.view.family.FamilyView;
import blue.koenig.kingsfamilylibrary.view.family.LoginHandler;
import blue.koenig.kingsfinances.R;
import blue.koenig.kingsfinances.model.calculation.AssetsCalculator;
import blue.koenig.kingsfinances.model.calculation.AssetsCalculatorService;
import blue.koenig.kingsfinances.model.calculation.DebtsCalculator;
import blue.koenig.kingsfinances.model.calculation.StatisticEntry;
import blue.koenig.kingsfinances.model.calculation.StatisticsCalculatorService;
import blue.koenig.kingsfinances.model.database.FinanceDatabase;
import blue.koenig.kingsfinances.view.FinanceNullView;
import blue.koenig.kingsfinances.view.FinanceView;
import blue.koenig.kingsfinances.view.NullPendingView;
import blue.koenig.kingsfinances.view.PendingView;

/**
 * Created by Thomas on 18.10.2017.
 */

public class FinanceModel extends FamilyModel implements FinanceCategoryService.CategoryServiceListener {

    private static final String DEBTS = "DEBTS";
    private static final String ASSETS = "ASSETS";
    private final FinanceUserService userService;
    private DebtsCalculator debtsCalculator;
    private FinanceDatabase database;
    private FinanceCategoryService categoryService;
    private PendingView pendingView;
    private int succesMessages;
    private List<Expenses> allExpenses;
    private AssetsCalculator assetsCalculator;

    public FinanceModel(ServerConnection connection, Context context, LoginHandler handler) {
        super(connection, context, handler);
        pendingView = new NullPendingView();
        categoryService = new FinanceCategoryService();
        categoryService.setListener(this);
        userService = new FinanceUserService(loginHandler.getMembers());

        try {
            database = new FinanceDatabase(context, userService);
            categoryService.update(database.getAllCategorys());
            debtsCalculator = new DebtsCalculator(Period.months(1), database.getExpensesTable(), new StatisticsCalculatorService() {
                @Override
                public List<StatisticEntry> getSavedSortedDebts() {
                    ByteBuffer buffer = FamilyConfig.getBytesFromConfig(context, DEBTS);
                    if (buffer == null) return new ArrayList<>();

                    int size = buffer.getInt();
                    List<StatisticEntry> debts = new ArrayList<>(size);
                    for (int i = 0; i < size; i++) {
                        debts.add(new StatisticEntry(buffer));
                    }

                    return debts;
                }

                @Override
                public void saveDebts(List<StatisticEntry> statisticEntryList) {
                    ByteBuffer buffer = ByteBuffer.allocate(Byteable.getBigListLength(statisticEntryList));
                    Byteable.writeBigList(statisticEntryList, buffer);
                    FamilyConfig.saveBytes(context, buffer.array(), DEBTS);
                }
            });
            assetsCalculator = new AssetsCalculator(Period.months(1), database.getBankAccountTable(), new AssetsCalculatorService() {
                @Override
                public Map<BankAccount, List<StatisticEntry>> getAllBankAccountStatistics() {
                    ByteBuffer buffer = FamilyConfig.getBytesFromConfig(context, ASSETS);
                    if (buffer == null) return new HashMap<>();

                    int size = buffer.getInt();
                    Map<BankAccount, List<StatisticEntry>> listMap = new HashMap<>(size);
                    for (int i = 0; i < size; i++) {
                        BankAccount bankAccount = new BankAccount(buffer);
                        int entries = buffer.getInt();
                        List<StatisticEntry> statistics = new ArrayList<>(entries);
                        for (int j = 0; j < entries; j++) {
                            statistics.add(new StatisticEntry(buffer));
                        }

                        listMap.put(bankAccount, statistics);
                    }

                    return listMap;
                }

                @Override
                public DateTime getStartDate() {
                    // TODO: preferences or somewhere
                    return new DateTime(2015, 1, 1, 0, 0);
                }

                @Override
                public DateTime getEndDate() {
                    return DateTime.now();
                }

                @Override
                public void save(Map<BankAccount, List<StatisticEntry>> statisticEntryLists) {
                    int size = 4;
                    for (BankAccount bankAccount : statisticEntryLists.keySet()) {
                        size += bankAccount.getByteLength();
                        size += Byteable.getBigListLength(statisticEntryLists.get(bankAccount));
                    }

                    ByteBuffer buffer = ByteBuffer.allocate(size);
                    buffer.putInt(statisticEntryLists.size());
                    for (BankAccount bankAccount : statisticEntryLists.keySet()) {
                        bankAccount.writeBytes(buffer);
                        Byteable.writeBigList(statisticEntryLists.get(bankAccount), buffer);
                    }

                    FamilyConfig.saveBytes(context, buffer.array(), ASSETS);
                }
            });
        } catch (SQLException e) {
            logger.error("Couldn't create database: " + e.getMessage());
        }
    }

    public static void update(FinanceDatabase database, List<DatabaseItem> items) throws SQLException {
        if (items.size() == 0) return;
        ItemType itemType = ItemType.fromItem(items.get(0).getItem());

        switch (itemType) {
            case EXPENSES:
                database.updateExpensesChanges(items);
                break;
            case STANDING_ORDER:
                database.updateStandingOrderChanges(items);
                break;
            case CATEGORY:
                database.updateCategoryChanges(items);
                break;
            case BANKACCOUNT:
                database.updateBankAccountChanges(items);
                break;
        }
    }

    @Override
    public void start() {
        logger.info("Start");
        List<User> members = loginHandler.getMembers();
        if (members != null) {
            updateFamilymembers(members);
        }

        List<PendingOperation> operations = database.getNonConfirmedPendingOperations();
        clearAUDSuccesMessages();
        for (PendingOperation operation : operations) {
            connection.sendFamilyMessage(new AUDMessage(Component.FINANCE,operation.getOperation()));
        }

        // ask for update after sent my changes!
        waitForSuccessMessages(operations.size());
        askForAllUpdates();
    }

    private void waitForSuccessMessages(int size) {
        int i = 0;
        int timeOutInS = 2;
        int intervall = 100;
        try {
            while (i < 1000 * timeOutInS / intervall) {
            if (succesMessages >= size) return;
            Thread.sleep(intervall);
            i++;
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void clearAUDSuccesMessages() {
        succesMessages = 0;
    }

    private void askForAllUpdates() {
        askForUpdates(ItemType.EXPENSES);
        askForUpdates(ItemType.CATEGORY);
        askForUpdates(ItemType.STANDING_ORDER);
        askForUpdates(ItemType.BANKACCOUNT);
    }

    private void askForUpdates(ItemType itemType) {
        sendMessageToServer(new AskForUpdatesMessage(Component.FINANCE, FamilyConfig.getLastSyncDate(context, itemType.name()) , itemType));
    }

    @Override
    protected void updateFamilymembers(List<User> members) {
        logger.info("Setting family members...");
        getFinanceView().setFamilyMembers(members);
        userService.setUser(members);
    }

    @Override
    public void onReceiveFinanceMessage(FamilyMessage message) {
            switch (message.getName()) {
                case UpdatesMessage.NAME:
                    UpdatesMessage updatesMessage = (UpdatesMessage) message;
                    update(updatesMessage.getItems());
                    break;
                default:
                    logger.error("Unknown Message: " + message.getName());
            }
    }

    private void update(List<DatabaseItem> items) {
        try {
            update(database, items);
            if (items.size() == 0) return;
            ItemType itemType = ItemType.fromItem(items.get(0).getItem());

            switch (itemType) {
                case EXPENSES:
                    updateAllExpensesAndDebts();
                    List<Expenses> expensesList = new ArrayList<>(items.size());
                    for (DatabaseItem item : items) {
                        expensesList.add((Expenses) item.getItem());
                    }

                    break;
                case STANDING_ORDER:
                    updateAllStandingOrders();
                    break;
                case CATEGORY:
                    updateAllCategorys();
                    break;
                case BANKACCOUNT:
                    updateAllBankAccountsAndAssets();
                    break;
            }

            FamilyConfig.saveLastSyncDate(DateTime.now(), context, itemType.name());

        } catch (SQLException ex) {
            logger.error("Error while updating: " + ex.getMessage());
        }
    }


    private void updateAllCategorys() {
        try {
            List<Category> categories = database.getAllCategorys();
            categoryService.update(categories);
        } catch (SQLException e) {
            logger.error("Error getting categoires: " + e.getMessage());
        }
    }

    private void updateAllStandingOrders() {
        try {
            List<StandingOrder> standingOrders = database.getAllStandingOrders();
            getFinanceView().showStandingOrders(standingOrders);
        } catch (SQLException e) {
            logger.error("Error getting categoires: " + e.getMessage());
        }
    }

    private FinanceView getFinanceView() {
        return (FinanceView) view;
    }

    public void onTabSelected(int position) {
        logger.info("OnTabSelected: " + position);
        if (position == 1) {
            pendingView.update(getPendingOperations());
        }
    }

    @Override
    protected void processFinanceCommand(String[] words) {
        try {

            switch (words[0]) {
                case FinanceTextMessages.GET_ALL_EXPENSES_FAIL:
                    view.showText(R.string.getExpensesFail);
                    break;
                case FinanceTextMessages.AUD_SUCCESS:
                    succesMessages++;
                    database.setPendingOperation(PendingStatus.CONFIRMED, words[1]);
                    pendingView.update(getPendingOperations());
                    break;
                case FinanceTextMessages.AUD_FAIL:
                    database.setPendingOperation(PendingStatus.ERROR, words[1]);
                    pendingView.update(getPendingOperations());
                    break;

                default:
                    logger.info("Unknown command");
                    view.showText(words[0]);
            }
        } catch (Exception e) {
            logger.error("Error while processing finance command: " + e.getMessage());
        }
    }

    public FamilyView createNullView() {
        return new FinanceNullView();
    }

    public void deleteExpenses(Expenses expenses) {
        makeDeleteOperation(expenses);

        try {
            database.deleteExpenses(expenses);
            updateAllExpensesAndDebts();
        } catch (SQLException e) {
            logger.error("Error while deleting expenses");
        }
    }


    private void makeUpdateOperation(Item item) {
        makeOperation(Operator.UPDATE, item);
    }

    private void makeAddOperation(Item item) {
        makeOperation(Operator.ADD, item);
    }

    private void makeDeleteOperation(Item item) {
        makeOperation(Operator.DELETE, item);
    }
    private void makeOperation(Operator operator, Item item) {
        Operation operation = new Operation(operator, item);
        try {
            database.addPendingOperation(new PendingOperation(operation, PendingStatus.PENDING, DateTime.now()));
        } catch (SQLException e) {
            logger.error("Error on operation: " + e.getMessage());
        }

        connection.sendFamilyMessage(new AUDMessage(Component.FINANCE, operation));
        pendingView.update(getPendingOperations());
    }

    public void editExpenses(Expenses expenses) {
        makeUpdateOperation(expenses);
        try {
            database.updateExpenses(expenses);
            updateAllExpensesAndDebts();
        } catch (SQLException e) {
            logger.error("Error while updating expenses: " + e.getMessage());
        }
    }

    public CategoryService getCategoryService() {
        return categoryService;
    }

    public void addExpenses(Expenses expenses) {
        try {
            makeAddOperation(expenses);

            database.addExpenses(expenses);
            updateAllExpensesAndDebts();
        } catch (SQLException e) {
            logger.error("Error while adding expenses: " + e.getMessage());
        }
    }

    private void updateAllExpensesAndDebts() throws SQLException {
        allExpenses = database.getAllExpenses();
        getFinanceView().showExpenses(allExpenses);
        getFinanceView().updateDebts(debtsCalculator.getEntrys());
    }

    private void updateAllBankAccountsAndAssets() throws SQLException {
        getFinanceView().updateBankAccounts(database.getAllBankAccounts());
        getFinanceView().updateAssets(getAllAssets());
    }

    public List<PendingOperation> getPendingOperations() {
        try {
            return database.getAllPendingOperation();
        } catch (SQLException e) {
            logger.error("Couldn't get pending operations: " + e.getMessage());
        }

        return new ArrayList<>();
    }

    public void deletePending(PendingOperation operation) {
        try {
            database.deletePendingOperation(operation.getId());
            // TODO: revert database(memorize last state) or ask database for old state
            pendingView.update(getPendingOperations());
        } catch (SQLException e) {
            logger.error("Exception deleting operation: " + e.getMessage());
        }
    }

    public void sendPending(PendingOperation pendingOperation) {
        connection.sendFamilyMessage(new AUDMessage(Component.FINANCE, pendingOperation.getOperation()));
    }

    public void attachPendingView(PendingView pendingFragment) {
        this.pendingView = pendingFragment;
    }

    public void detachPendingView() {
        this.pendingView = new NullPendingView();
    }


    public List<Expenses> getExpenses() {
        // TODO: dauert viel zu lange, im Hintergrund ausführen oder hier zwischenspeichern und nur ab und zu aktualisieren
        if (allExpenses != null) return allExpenses;
        try {
            allExpenses = database.getAllExpenses();
            return allExpenses;
        } catch (SQLException e) {
            logger.error("Couldn't get expenses: " + e.getMessage());
        }


        return new ArrayList<>();
    }

    @Override
    public void addMainCategory(String newCategory) {
        try {
            Category category = new Category(newCategory);
            makeAddOperation(category);
            database.addCategory(category);
        } catch (SQLException e) {
            logger.error("Error adding new Category");
        }
    }

    @Override
    public void addSubCategory(String mainCategory, String newCategory) {
            try {
                Category category = database.getCategory(mainCategory);
                if (category != null) {
                    category.addSub(newCategory);
                    makeUpdateOperation(category);
                    database.updateCategory(category);
                } else {
                    logger.error("Couldn't find main category to add subcategory: " + mainCategory);
                }

            } catch (SQLException e) {
                logger.error("Error adding new subcategory");
            }
    }

    public List<StandingOrder> getStandingOrders()  {
        try {
            return database.getAllStandingOrders();
        } catch (SQLException e) {
            logger.error("Error getting standing orders");
            return new ArrayList<>();
        }
    }

    public void addStandingOrder(StandingOrder standingOrder) {
        try {
            makeAddOperation(standingOrder);
            database.addStandingOrder(standingOrder);
            updateAllStandingOrders();
        } catch (SQLException e) {
            logger.error("Error while adding standingOrder: " + e.getMessage());
        }
    }

    public void editStandingOrder(StandingOrder standingOrder) {
        makeUpdateOperation(standingOrder);
        try {
            database.updateStandingOrder(standingOrder);
            updateAllStandingOrders();
        } catch (SQLException e) {
            logger.error("Error while updating standingOrder: " + e.getMessage());
        }
    }

    public void deleteStandingOrder(StandingOrder standingOrder) {
        makeDeleteOperation(standingOrder);
        try {
            database.deleteStandingOrder(standingOrder);
            updateAllStandingOrders();
        } catch (SQLException e) {
            logger.error("Error while deleting standingOrder");
        }
    }

    public List<BankAccount> getBankAccounts() {
        try {
            return database.getAllBankAccounts();
        } catch (SQLException e) {
            logger.error("Couldn't get bankaccounts: " + e.getMessage());
            return new ArrayList<>();
        }
    }

    public void deleteBankAccount(BankAccount account) {
        makeDeleteOperation(account);
        try {
            database.deleteBankAccount(account);
            updateAllBankAccountsAndAssets();
        } catch (SQLException e) {
            logger.error("Error while deleting standingOrder");
        }
    }


    public void deleteBalance(BankAccount account, Balance balance) {
        makeUpdateOperation(account);
        try {
            database.deleteBalance(account, balance);
            updateAllBankAccountsAndAssets();
        } catch (SQLException e) {
            logger.error("Couldn't delete balance: " + e.getMessage());
        }
    }

    public void addBalance(BankAccount account, Balance balance) {
        makeUpdateOperation(account);
        try {
            database.addBalance(account, balance);
            updateAllBankAccountsAndAssets();
        } catch (SQLException e) {
            logger.error("Couldn't add balance: " + e.getMessage());
        }
    }

    public void addBankAccount(BankAccount bankAccount) {
        makeAddOperation(bankAccount);
        try {
            database.addBankAccount(bankAccount);
            updateAllBankAccountsAndAssets();
        } catch (SQLException e) {
            logger.error("Couldn't add bankAccount: " + e.getMessage());
        }
    }

    public List<StatisticEntry> getDebts() {
        List<StatisticEntry> debts = debtsCalculator.getEntrys();
        if (debts.size() == 0) {
            logger.warn("Recalculating all debts!");
            //return debtsCalculator.recalculateAll();
        }

        return debts;
    }

    public int getColorFor(User user) {
        // TODO: make preferences
        if (user.getName().equals("Thomas")) return Color.BLUE;
        else return Color.MAGENTA;
    }

    public List<StatisticEntry> getAllAssets() {
        return assetsCalculator.getEntrysForAll();
    }
}

File: app/src/main/java/blue/koenig/kingsfinances/model/FinanceModel.kt
package blue.koenig.kingsfinances.model

import android.content.Context
import android.graphics.Color
import blue.koenig.kingsfamilylibrary.model.FamilyConfig
import blue.koenig.kingsfamilylibrary.model.communication.ServerConnection
import blue.koenig.kingsfamilylibrary.model.family.FamilyModel
import blue.koenig.kingsfamilylibrary.view.family.FamilyView
import blue.koenig.kingsfamilylibrary.view.family.LoginHandler
import blue.koenig.kingsfinances.R
import blue.koenig.kingsfinances.features.category_statistics.CategoryCalculator
import blue.koenig.kingsfinances.features.standing_orders.StandingOrderExecutor
import blue.koenig.kingsfinances.features.statistics.AssetsCalculator
import blue.koenig.kingsfinances.model.calculation.DebtsCalculator
import blue.koenig.kingsfinances.model.calculation.IncomeCalculator
import blue.koenig.kingsfinances.model.calculation.StatisticEntry
import blue.koenig.kingsfinances.model.database.FinanceDatabase
import blue.koenig.kingsfinances.view.FinanceNullView
import blue.koenig.kingsfinances.view.FinanceView
import blue.koenig.kingsfinances.view.NullPendingView
import blue.koenig.kingsfinances.view.PendingView
import com.koenig.commonModel.*
import com.koenig.commonModel.database.DatabaseItem
import com.koenig.commonModel.finance.Balance
import com.koenig.commonModel.finance.BankAccount
import com.koenig.commonModel.finance.Expenses
import com.koenig.commonModel.finance.StandingOrder
import com.koenig.communication.messages.AUDMessage
import com.koenig.communication.messages.AskForUpdatesMessage
import com.koenig.communication.messages.FamilyMessage
import com.koenig.communication.messages.UpdatesMessage
import com.koenig.communication.messages.finance.FinanceTextMessages
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.joda.time.DateTime
import org.slf4j.LoggerFactory
import java.sql.SQLException
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Thomas on 18.10.2017.
 */

class FinanceModel// TODO: Income Calculator as input that its tracking all changes from beginning. Put somewhere else but make sure it is tracking from startLoggedIn!
(connection: ServerConnection, context: Context, handler: LoginHandler, internal var database: FinanceDatabase, internal var userService: FinanceUserService, private val assetsCalculator: AssetsCalculator, incomeCalculator: IncomeCalculator, categoryCalculator: CategoryCalculator, val standingOrderExecutor: StandingOrderExecutor, val debtsCalculator: DebtsCalculator) : FamilyModel(connection, context, handler), FinanceCategoryService.CategoryServiceListener {

    private val categoryService: FinanceCategoryService
    private var pendingView: PendingView? = null
    private var succesMessages: Int = 0
    private var allExpenses: List<Expenses>? = null
    private var updateMessages: Int = 0
    private val financeView: FinanceView
        get() = view as FinanceView

    override fun start() {
        logger.info("Start")
        // start executing standing orders after while because at the beginning is so much work done
        Observable.timer(20, TimeUnit.SECONDS).observeOn(Schedulers.computation()).subscribe {
            logger.info("Executing standing orders...")
            standingOrderExecutor.executeForAll()
            if (standingOrderExecutor.consistencyCheck()) {
                logger.info("Consistency Check for standing orders passed")
            } else {
                logger.error("Consistency Check for standing orders failed")
                view.showText("Consistency Check for standing orders failed")
            }
        }
    }

    val pendingOperations: List<PendingOperation<out Item>>
        get() {
            try {
                return database.allPendingOperation
            } catch (e: SQLException) {
                logger.error("Couldn't get pending operations: " + e.message)
            }

            return ArrayList()
        }


    // TODO: dauert viel zu lange, im Hintergrund ausführen oder hier zwischenspeichern und nur ab und zu aktualisieren, oder nur die ersten 100 zurückliefern
    val expenses: List<Expenses>?
        get() {
            //if (allExpenses != null) return allExpenses
            try {
                allExpenses = database.allExpenses
                return allExpenses
            } catch (e: SQLException) {
                logger.error("Couldn't get expenses: " + e.message)
            }


            return ArrayList()
        }

    val standingOrders: List<StandingOrder>
        get() {
            try {
                return database.allStandingOrders
            } catch (e: SQLException) {
                logger.error("Error getting standing orders")
                return ArrayList()
            }

        }

    val bankAccounts: List<BankAccount>
        get() {
            try {
                return database.allBankAccounts
            } catch (e: SQLException) {
                logger.error("Couldn't get bankaccounts: " + e.message)
                return ArrayList()
            }

        }

    //return debtsCalculator.recalculateAll();
    val debts: List<StatisticEntry>
        get() {
            val debts = debtsCalculator.entrys
            if (debts.size == 0) {
                logger.warn("Recalculating all debts!")
            }

            return debts
        }

    val allAssets: List<StatisticEntry>
        get() = assetsCalculator.entrysForAll

    init {
        pendingView = NullPendingView()
        categoryService = FinanceCategoryService()
        categoryService.setListener(this)

        try {
            categoryService.update(database.allCategorys)

        } catch (e: SQLException) {
            logger.error("Couldn't create database: " + e.message)
        }

    }

    public override fun startLoggedIn() {
        logger.info("Start")
        val members = loginHandler.members
        if (members != null) {
            updateFamilymembers(members)
        }

        // TODO: send only last edit from same item
        val operations = database.nonConfirmedPendingOperations
        clearAUDSuccesMessages()
        for (operation in operations) {
            connection.sendFamilyMessage(AUDMessage(Component.FINANCE, operation.operation))
        }

        // ask for update after sent my changes!
        waitForSuccessMessages(operations.size)
        updateMessages = 0
        askForAllUpdates()
        //waitForFinishUpdates()

    }

    private fun waitForFinishUpdates() {
        var i = 0
        val timeOutInS = 10
        val intervall = 100
        while (i < 1000 * timeOutInS / intervall) {
            if (updateMessages >= 5) return
            Thread.sleep(intervall.toLong())
            i++
        }

    }

    private fun waitForSuccessMessages(size: Int) {
        var i = 0
        val timeOutInS = 2
        val intervall = 100
        try {
            while (i < 1000 * timeOutInS / intervall) {
                if (succesMessages >= size) return
                Thread.sleep(intervall.toLong())
                i++
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    private fun clearAUDSuccesMessages() {
        succesMessages = 0
    }

    private fun askForAllUpdates() {
        askForUpdates(ItemType.EXPENSES)
        askForUpdates(ItemType.CATEGORY)
        askForUpdates(ItemType.STANDING_ORDER)
        askForUpdates(ItemType.BANKACCOUNT)
        askForUpdates(ItemType.GOAL)
    }

    private fun askForUpdates(itemType: ItemType) {
        sendMessageToServer(AskForUpdatesMessage(Component.FINANCE, FamilyConfig.getLastSyncDate(context, itemType.name), itemType))
    }

    override fun updateFamilymembers(members: List<User>) {
        logger.info("Setting family members...")
        financeView.setFamilyMembers(members)
        userService.setUser(members)
    }


    public override fun onReceiveFinanceMessage(message: FamilyMessage) {
        when (message.name) {
            UpdatesMessage.NAME -> {
                val updatesMessage = message as UpdatesMessage<*>
                update(updatesMessage.items)
                updateMessages++
            }
            else -> logger.error("Unknown Message: " + message.name)
        }
    }

    private fun update(items: List<DatabaseItem<*>>) {
        try {
            update(database, items)
            if (items.isEmpty()) return
            val itemType = ItemType.fromItem(items[0].item)

            when (itemType) {
                ItemType.EXPENSES -> updateAllExpensesAndDebts()
                ItemType.STANDING_ORDER -> updateAllStandingOrders()
                ItemType.CATEGORY -> updateAllCategorys()
                ItemType.BANKACCOUNT -> updateAllBankAccountsAndAssets()
                ItemType.GOAL -> updateGoals()
                else -> logger.error("Unknown item type $itemType")
            }

            FamilyConfig.saveLastSyncDate(DateTime.now(), context, itemType.name)

        } catch (ex: SQLException) {
            logger.error("Error while updating: " + ex.message)
        }

    }

    private fun updateGoals() {
        // TODO: update CategoryStatisticsFragment/Presenter --> It should receive messages on its own
    }


    private fun updateAllCategorys() {
        try {
            val categories = database.allCategorys
            categoryService.update(categories)
        } catch (e: SQLException) {
            logger.error("Error getting categoires: " + e.message)
        }

    }

    private fun updateAllStandingOrders() {
        try {
            val standingOrders = database.allStandingOrders
            financeView.showStandingOrders(standingOrders)
        } catch (e: SQLException) {
            logger.error("Error getting categoires: " + e.message)
        }

    }

    fun onTabSelected(position: Int) {
        logger.info("OnTabSelected: " + position)
        if (position == 1) {
            pendingView!!.update(pendingOperations)
        }
    }

    override fun processFinanceCommand(words: Array<String>) {
        try {

            when (words[0]) {
                FinanceTextMessages.GET_ALL_EXPENSES_FAIL -> view.showText(R.string.getExpensesFail)
                FinanceTextMessages.AUD_SUCCESS -> {
                    succesMessages++
                    database.setPendingOperation(PendingStatus.CONFIRMED, words[1])
                    pendingView!!.update(pendingOperations)
                }
                FinanceTextMessages.AUD_FAIL -> {
                    database.setPendingOperation(PendingStatus.ERROR, words[1])
                    pendingView!!.update(pendingOperations)
                }

                else -> {
                    logger.info("Unknown command")
                    view.showText(words[0])
                }
            }
        } catch (e: Exception) {
            logger.error("Error while processing finance command: " + e.message)
        }

    }

    public override fun createNullView(): FamilyView {
        return FinanceNullView()
    }

    fun deleteExpenses(expenses: Expenses) {
        makeDeleteOperation(expenses)

        try {
            database.deleteExpenses(expenses)
            updateAllExpensesAndDebts()
        } catch (e: SQLException) {
            logger.error("Error while deleting expenses")
        }

    }


    private fun makeUpdateOperation(item: Item) {
        makeOperation(Operator.UPDATE, item)
    }

    private fun makeAddOperation(item: Item) {
        makeOperation(Operator.ADD, item)
    }

    private fun makeDeleteOperation(item: Item) {
        makeOperation(Operator.DELETE, item)
    }

    private fun makeOperation(operator: Operator, item: Item) {
        val operation = Operation(operator, item)
        try {
            database.addPendingOperation(PendingOperation(operation, PendingStatus.PENDING, DateTime.now()))
        } catch (e: SQLException) {
            logger.error("Error on operation: " + e.message)
        }

        connection.sendFamilyMessage(AUDMessage(Component.FINANCE, operation))
        pendingView!!.update(pendingOperations)
    }

    fun editExpenses(expenses: Expenses) {
        makeUpdateOperation(expenses)
        try {
            database.updateExpenses(expenses)
            updateAllExpensesAndDebts()
        } catch (e: SQLException) {
            logger.error("Error while updating expenses: " + e.message)
        }

    }

    fun getCategoryService(): CategoryService {
        return categoryService
    }

    fun addExpenses(expenses: Expenses) {
        try {
            makeAddOperation(expenses)

            database.addExpenses(expenses)
            updateAllExpensesAndDebts()
        } catch (e: SQLException) {
            logger.error("Error while adding expenses: " + e.message)
        }

    }

    @Throws(SQLException::class)
    private fun updateAllExpensesAndDebts() {
        // TODO: notify expenses changed through database
    }

    @Throws(SQLException::class)
    private fun updateAllBankAccountsAndAssets() {
        financeView.updateBankAccounts(database.allBankAccounts)
        financeView.updateAssets(allAssets)
    }

    fun deletePending(operation: PendingOperation<Item>) {
        try {
            database.deletePendingOperation(operation.id)
            // TODO: revert database(memorize last state) or ask database for old state
            pendingView!!.update(pendingOperations)
        } catch (e: SQLException) {
            logger.error("Exception deleting operation: " + e.message)
        }

    }

    fun sendPending(pendingOperation: PendingOperation<Item>) {
        connection.sendFamilyMessage(AUDMessage(Component.FINANCE, pendingOperation.operation))
    }

    fun attachPendingView(pendingFragment: PendingView) {
        this.pendingView = pendingFragment
    }

    fun detachPendingView() {
        this.pendingView = NullPendingView()
    }

    override fun addMainCategory(newCategory: String) {
        try {
            val category = Category(newCategory)
            makeAddOperation(category)
            database.addCategory(category)
        } catch (e: SQLException) {
            logger.error("Error adding new Category")
        }

    }

    override fun addSubCategory(mainCategory: String, newCategory: String) {
        try {
            val category = database.getCategory(mainCategory)
            if (category != null) {
                category.addSub(newCategory)
                makeUpdateOperation(category)
                database.updateCategory(category)
            } else {
                logger.error("Couldn't find main category to add subcategory: " + mainCategory)
            }

        } catch (e: SQLException) {
            logger.error("Error adding new subcategory")
        }

    }

    fun addStandingOrder(standingOrder: StandingOrder) {
        try {
            makeAddOperation(standingOrder)
            database.addStandingOrder(standingOrder)
            updateAllStandingOrders()
        } catch (e: SQLException) {
            logger.error("Error while adding standingOrder: " + e.message)
        }

    }

    fun editStandingOrder(standingOrder: StandingOrder) {
        makeUpdateOperation(standingOrder)
        try {
            database.updateStandingOrder(standingOrder)
            updateAllStandingOrders()
        } catch (e: SQLException) {
            logger.error("Error while updating standingOrder: " + e.message)
        }

    }

    fun deleteStandingOrder(standingOrder: StandingOrder) {
        makeDeleteOperation(standingOrder)
        try {
            database.deleteStandingOrder(standingOrder)
            updateAllStandingOrders()
        } catch (e: SQLException) {
            logger.error("Error while deleting standingOrder")
        }

    }

    fun deleteBankAccount(account: BankAccount) {
        makeDeleteOperation(account)
        try {
            database.deleteBankAccount(account)
            updateAllBankAccountsAndAssets()
        } catch (e: SQLException) {
            logger.error("Error while deleting standingOrder")
        }

    }


    fun deleteBalance(account: BankAccount, balance: Balance) {
        makeUpdateOperation(account)
        try {
            database.deleteBalance(account, balance)
            updateAllBankAccountsAndAssets()
        } catch (e: SQLException) {
            logger.error("Couldn't delete balance: " + e.message)
        }

    }

    fun addBalance(account: BankAccount, balance: Balance) {
        makeUpdateOperation(account)
        try {
            database.addBalance(account, balance)
            updateAllBankAccountsAndAssets()
        } catch (e: SQLException) {
            logger.error("Couldn't add balance: " + e.message)
        }

    }

    fun addBankAccount(bankAccount: BankAccount) {
        makeAddOperation(bankAccount)
        try {
            database.addBankAccount(bankAccount)
            updateAllBankAccountsAndAssets()
        } catch (e: SQLException) {
            logger.error("Couldn't add bankAccount: " + e.message)
        }

    }

    fun getColorFor(user: User): Int {
        // TODO: make preferences
        return if (user.name == "Thomas")
            Color.BLUE
        else
            Color.MAGENTA
    }

    companion object {

        private val clogger = LoggerFactory.getLogger("FinanceModel")

        @Suppress("UNCHECKED_CAST")
        @Throws(SQLException::class)
        fun update(database: FinanceDatabase, items: List<DatabaseItem<*>>) {
            if (items.isEmpty()) return
            val itemType = ItemType.fromItem(items[0].item)

            when (itemType) {
                ItemType.EXPENSES -> database.updateExpensesChanges(items as List<DatabaseItem<Expenses>>)
                ItemType.STANDING_ORDER -> database.updateStandingOrderChanges(items as List<DatabaseItem<StandingOrder>>)
                ItemType.CATEGORY -> database.updateCategoryChanges(items as List<DatabaseItem<Category>>)
                ItemType.BANKACCOUNT -> database.updateBankAccountChanges(items as List<DatabaseItem<BankAccount>>)
                ItemType.GOAL -> database.updateGoalChanges(items as List<DatabaseItem<Goal>>)
                else -> clogger.error("Unknown ItemType $itemType")
            }
        }
    }

}
